sdks/python/apache_beam/runners/common.py

```diff
@@ -51,117 +51,192 @@ def receive(self, windowed_value):
+
+  def __init__(self, signature):
+    self.signature = signature
+
+  @staticmethod
+  def create_invoker(
+      signature, context, side_inputs, input_args,
+      input_kwargs):
+    """ Creates a new DoFnInvoker based on given arguments.
+
+    Args:
+        signature: A DoFnSignature for the DoFn being invoked.
+        context: Context to be used when invoking the DoFn (deprecated).
+        side_inputs: side inputs to be used when invoking th process method.
+        input_args: arguments to be used when invoking the process method
+        input_kwargs: kwargs to be used when invoking the process method.
+    """
+    default_arg_values = signature.process_method.defaults
+    use_simple_invoker = (
+        not side_inputs and not input_args and not input_kwargs and
+        not default_arg_values)
+    if use_simple_invoker:
+      return SimpleInvoker(signature)
     else:
-      assert context is not None
-      self.context = context
+      return PerWindowInvoker(
+          signature, context, side_inputs, input_args, input_kwargs)
 
-    class ArgPlaceholder(object):
-      def __init__(self, placeholder):
-        self.placeholder = placeholder
+  def invoke_process(self, element, process_output_fn):
+    raise NotImplementedError
 
-    # Stash values for use in dofn_process.
-    self.side_inputs = side_inputs
-    self.has_windowed_inputs = not all(
-        si.is_globally_windowed() for si in self.side_inputs)
-
-    self.args = args if args else []
-    self.kwargs = kwargs if kwargs else {}
-    self.dofn = fn
-    self.dofn_process = fn.process
-
-    arguments, _, _, defaults = self.dofn.get_function_arguments('process')
-    defaults = defaults if defaults else []
-    self_in_args = int(self.dofn.is_process_bounded())
-
-    self.use_simple_invoker = (
-        not side_inputs and not args and not kwargs and not defaults)
-    if self.use_simple_invoker:
-      # As we're using the simple invoker we don't need to compute placeholders
-      return
+  def invoke_start_bundle(self, process_output_fn):
+    defaults = self.signature.start_bundle_method.defaults
+    args = [self.context if d == core.DoFn.ContextParam else d
+            for d in defaults]
+    process_output_fn(None, self.signature.start_bundle_method.call(args, {}))
+
+  def invoke_finish_bundle(self, process_output_fn):
+    defaults = self.signature.finish_bundle_method.defaults
+    args = [self.context if d == core.DoFn.ContextParam else d
+            for d in defaults]
+    process_output_fn(None, self.signature.finish_bundle_method.call(args, {}))
 
-    self.has_windowed_inputs = (self.has_windowed_inputs or
-                                core.DoFn.WindowParam in defaults)
+
+class SimpleInvoker(DoFnInvoker):
+  """An invoker that processes elements ignoring windowing information."""
+
+  def invoke_process(self, element, process_output_fn):
+    process_output_fn(element, self.signature.process_method.call(
```

Comment:
I could get rid of the extra per-element overhead you mentioned above by caching the process method within invokers. There is still an extra method call compared to the current implementation (calling a method within DoFnInvoker instead of a method within DoFnRunner). But this does not seem to be significant based on results of some benchmarks I ran. I'll post a comment with benchmark results.